File: articles/adult-development-and-the-psychology-of-life-stages/r/model_adult_development.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(dplyr)
  library(ggplot2)
  library(lme4)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- normalizePath(sub(file_arg, "", args[grep(file_arg, args)]))
root <- normalizePath(file.path(dirname(script_path), ".."), mustWork = TRUE)

data_path <- file.path(root, "data", "adult_development_life_stages_panel.csv")
outputs_dir <- file.path(root, "outputs")

if (!dir.exists(outputs_dir)) dir.create(outputs_dir, recursive = TRUE)
if (!file.exists(data_path)) stop("Run python/generate_adult_development_panel.py first.")

panel <- read.csv(data_path)

adjustment_model <- lmer(
  adjustment_score ~ time + current_relational_support +
    current_work_integration + current_health_burden +
    current_adaptive_resources + current_role_burden +
    institutional_support + community_stability +
    young_stage + midlife_stage + later_stage +
    current_relational_support:current_adaptive_resources +
    (1 + time | context_id/id),
  data = panel
)

burden_model <- lmer(
  adjustment_score ~ time + current_role_burden + current_health_burden +
    current_relational_support + current_adaptive_resources +
    current_work_integration + institutional_support +
    life_stage + (1 + time | context_id/id),
  data = panel
)

capture.output(
  list(
    adjustment_model = summary(adjustment_model),
    burden_model = summary(burden_model)
  ),
  file = file.path(outputs_dir, "r_adult_development_model_summary.txt")
)

trajectory <- panel |>
  group_by(time, life_stage) |>
  summarize(
    average_adjustment = mean(adjustment_score),
    standard_error = sd(adjustment_score) / sqrt(n()),
    lower = average_adjustment - 1.96 * standard_error,
    upper = average_adjustment + 1.96 * standard_error,
    .groups = "drop"
  )

write.csv(trajectory, file.path(outputs_dir, "r_adult_stage_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_adult_stage_trajectory.png"),
  plot = ggplot(trajectory, aes(x = time, y = average_adjustment, linetype = life_stage)) +
    geom_line(linewidth = 1) +
    geom_ribbon(aes(ymin = lower, ymax = upper, group = life_stage), alpha = 0.12) +
    labs(title = "Synthetic Adult Development Across Life Stages", x = "Time", y = "Average adjustment", linetype = "Life stage") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

support_burden <- panel |>
  group_by(time) |>
  summarize(
    average_support = mean(current_relational_support),
    average_work = mean(current_work_integration),
    average_health = mean(current_health_burden),
    average_resources = mean(current_adaptive_resources),
    average_role_burden = mean(current_role_burden),
    .groups = "drop"
  )

write.csv(support_burden, file.path(outputs_dir, "r_adult_support_burden_trajectory.csv"), row.names = FALSE)

ggsave(
  filename = file.path(outputs_dir, "r_adult_support_burden_trajectory.png"),
  plot = ggplot(support_burden, aes(x = time)) +
    geom_line(aes(y = average_support, linetype = "relational support"), linewidth = 1) +
    geom_line(aes(y = average_work, linetype = "work integration"), linewidth = 1) +
    geom_line(aes(y = average_health, linetype = "health burden"), linewidth = 1) +
    geom_line(aes(y = average_role_burden, linetype = "role burden"), linewidth = 1) +
    labs(title = "Synthetic Adult Support, Work, Health, and Role Burden", x = "Time", y = "Average index", linetype = "Measure") +
    theme_minimal(),
  width = 8,
  height = 5,
  dpi = 160
)

message("Wrote R adult development outputs.")

File: articles/adult-development-and-the-psychology-of-life-stages/python/model_adult_development.py
#!/usr/bin/env python3
"""Model synthetic adult development and the psychology of life stages."""

from __future__ import annotations

from pathlib import Path
import matplotlib.pyplot as plt
import pandas as pd
import statsmodels.formula.api as smf

ROOT = Path(__file__).resolve().parents[1]
DATA_PATH = ROOT / "data" / "adult_development_life_stages_panel.csv"
OUTPUTS_DIR = ROOT / "outputs"
OUTPUTS_DIR.mkdir(parents=True, exist_ok=True)


def main() -> None:
    if not DATA_PATH.exists():
        raise FileNotFoundError("Run python/generate_adult_development_panel.py first.")

    panel = pd.read_csv(DATA_PATH).sort_values(["id", "time"]).reset_index(drop=True)
    panel["lag_score"] = panel.groupby("id")["adjustment_score"].shift(1)
    model_data = panel.dropna(subset=["lag_score"]).copy()

    adjustment_model = smf.ols(
        """
        adjustment_score ~ lag_score + time + current_relational_support +
        current_work_integration + current_health_burden +
        current_adaptive_resources + current_role_burden +
        institutional_support + community_stability +
        young_stage + midlife_stage + later_stage +
        current_relational_support:current_adaptive_resources
        """,
        data=model_data,
    ).fit(cov_type="HC3")

    burden_model = smf.ols(
        """
        adjustment_score ~ time + current_role_burden + current_health_burden +
        current_relational_support + current_adaptive_resources +
        current_work_integration + institutional_support +
        C(life_stage)
        """,
        data=panel,
    ).fit(cov_type="HC3")

    with open(OUTPUTS_DIR / "python_adult_development_model_summary.txt", "w", encoding="utf-8") as f:
        f.write("DYNAMIC ADULT DEVELOPMENT MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(adjustment_model.summary().as_text())
        f.write("\n\nROLE BURDEN AND HEALTH BURDEN MODEL\n")
        f.write("=" * 80 + "\n")
        f.write(burden_model.summary().as_text())

    trajectory = panel.groupby(["time", "life_stage"], as_index=False).agg(
        average_adjustment=("adjustment_score", "mean"),
        standard_error=("adjustment_score", lambda x: x.std() / (len(x) ** 0.5)),
    )
    trajectory["lower"] = trajectory["average_adjustment"] - 1.96 * trajectory["standard_error"]
    trajectory["upper"] = trajectory["average_adjustment"] + 1.96 * trajectory["standard_error"]
    trajectory.to_csv(OUTPUTS_DIR / "python_adult_stage_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    for group_name, group in trajectory.groupby("life_stage"):
        plt.plot(group["time"], group["average_adjustment"], marker="o", label=str(group_name))
    plt.xlabel("Time")
    plt.ylabel("Average adjustment score")
    plt.title("Synthetic Adult Development Across Life Stages")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_adult_stage_trajectory.png", dpi=160)
    plt.close()

    support_burden = panel.groupby("time", as_index=False).agg(
        average_support=("current_relational_support", "mean"),
        average_work=("current_work_integration", "mean"),
        average_health=("current_health_burden", "mean"),
        average_resources=("current_adaptive_resources", "mean"),
        average_role_burden=("current_role_burden", "mean"),
    )
    support_burden.to_csv(OUTPUTS_DIR / "python_adult_support_burden_trajectory.csv", index=False)

    plt.figure(figsize=(8, 5))
    plt.plot(support_burden["time"], support_burden["average_support"], marker="o", label="relational support")
    plt.plot(support_burden["time"], support_burden["average_work"], marker="o", label="work integration")
    plt.plot(support_burden["time"], support_burden["average_health"], marker="o", label="health burden")
    plt.plot(support_burden["time"], support_burden["average_role_burden"], marker="o", label="role burden")
    plt.xlabel("Time")
    plt.ylabel("Average index")
    plt.title("Synthetic Adult Support, Work, Health, and Role Burden")
    plt.legend()
    plt.tight_layout()
    plt.savefig(OUTPUTS_DIR / "python_adult_support_burden_trajectory.png", dpi=160)
    plt.close()

    profile_summary = panel.groupby("adult_development_profile", as_index=False).agg(
        adults=("id", "nunique"),
        average_adjustment=("adjustment_score", "mean"),
        average_support=("current_relational_support", "mean"),
        average_work=("current_work_integration", "mean"),
        average_health=("current_health_burden", "mean"),
        average_resources=("current_adaptive_resources", "mean"),
        average_role_burden=("current_role_burden", "mean"),
    )
    profile_summary.to_csv(OUTPUTS_DIR / "python_adult_development_profiles.csv", index=False)

    print("Wrote Python adult development model outputs.")


if __name__ == "__main__":
    main()
